pymc3/distributions/distribution.py

```diff
@@ -198,29 +198,315 @@ class DensityDist(Distribution):
+                    mu = pm.Normal('mu', 0 , 1)
+                    normal_dist = pm.Normal.dist(mu, 1, shape=3)
+                    dens = pm.DensityDist(
+                        'density_dist',
+                        normal_dist.logp,
+                        observed=np.random.randn(100, 3),
+                        shape=3,
+                        random=normal_dist.random,
+                        wrap_random_with_dist_shape=False, # Is True by default
+                        check_shape_in_random=False, # Is True by default
+                    )
+                    prior = pm.sample_prior_predictive(10)['density_dist']
+                    # We get samples with an incorrect shape
+                    assert prior.shape != (10, 100, 3)
+    
+            One final word of caution. If you use a pymc3 distribution's random
+            method, you should stick with ``pymc3_size_interpretation=True``, or
+            you will get incorrectly shaped samples.
+
+
+            .. code-block:: python
+
+                with pm.Model():
+                    mu = pm.Normal('mu', 0 , 1)
+                    normal_dist = pm.Normal.dist(mu, 1, shape=3)
+                    dens = pm.DensityDist(
+                        'density_dist',
+                        normal_dist.logp,
+                        observed=np.random.randn(100, 3),
+                        shape=3,
+                        random=normal_dist.random,
+                        pymc3_size_interpretation=False, # Is True by default
+                    )
+                    prior = pm.sample_prior_predictive(10)['density_dist']
+                assert prior.shape != (10, 100, 3)
+                assert prior.shape == (10, 100, 3, 3)
+    
+            If you use callables that work with ``scipy.stats`` rvs, you should
+            set ``pymc3_size_interpretation=False``.
+
+
+            .. code-block:: python
+
+                with pm.Model():
+                    mu = pm.Normal('mu', 0 , 1)
+                    normal_dist = pm.Normal.dist(mu, 1, shape=3)
+                    dens = pm.DensityDist(
+                        'density_dist',
+                        normal_dist.logp,
+                        observed=np.random.randn(100, 3),
+                        shape=3,
+                        random=stats.norm.rvs,
+                        pymc3_size_interpretation=False, # Is True by default
+                    )
+                    prior = pm.sample_prior_predictive(10)['density_dist']
+                assert prior.shape == (10, 100, 3)
+
+        """
         if dtype is None:
             dtype = theano.config.floatX
         super().__init__(shape, dtype, testval, *args, **kwargs)
         self.logp = logp
         self.rand = random
+        self.wrap_random_with_dist_shape = wrap_random_with_dist_shape
+        self.check_shape_in_random = check_shape_in_random
+        self.pymc3_size_interpretation = pymc3_size_interpretation
```

Comment:
Should we really expose these option to user? I think it might be better in terms of API to just always do these check and raise error when the input function does not have the correct shape handling.
Granted that would place the burden to user to provide a rigid function, but on the other hand there are much less option exposed.

WDYT? @twiecki

Reply:
It is certainly not nice to have these extra parameters here. They only allow the users to go back to the old behavior. I don't know if this is a values enough reason to include them into the construction signature

Reply:
@junpenglao I think if we were to remove these we should just go go always using `wrap_random_with_dist_shape=True` (or the equivalent).  This might be backwards-incompatible, but the backwards way was wrong, as far as I can tell.

Reply:
I dont think we need to be backward compatible in this case if the old way was wrong/incomplete.

Reply:
I removed the `pymc3_size_interpretation` kwarg